remove final dependencies on tpraos

JaredCorduan opened the pull request:
This PR finishes the task of removing the dependency of `cardano-protocol-tpraos` in `cardano-ledger-shelley`. In fact, `cardano-protocol-tpraos` now depends on `cardano-ledger-shelley`.

- `Cardano/Ledger/Shelley/API/Protocol.hs` was moved to `Cardano/Protocol/TPraos/API.hs`
- the constraint `PraosCrypto` was removed from `ShelleyBasedEra`.
- `mkInitialShelleyLedgerView`, `GetLedgerView`, and `PraosCrypto` were moved to `Cardano.Protocol.TPraos.API`
- all the TPraos types in `Cardano/Ledger/Shelley/API/Types.hs` are removed

closes #2462

JaredCorduan commented on eras/shelley/impl/src/Cardano/Ledger/Shelley/API/Mempool.hs:
```diff
@@ -183,7 +183,25 @@ class
+    ToCBOR (PredicateFailure (Core.EraRule "DELEGS" (ShelleyEra c))),
+    ToCBOR (PredicateFailure (Core.EraRule "UTXOW" (ShelleyEra c)))
+  ) =>
+  ApplyTx (ShelleyEra c)
```
Do we like the `ApplyTx` remaining here, but changing the `PraosCrypto` constraint? I would love to understand why removing `PraosCrypto` forced all the constraints above, they do not seem to be covered by `PraosCrypto`.

nc6 replied:
Odd

JaredCorduan commented on libs/cardano-protocol-tpraos/src/Cardano/Protocol/TPraos/API.hs:
```diff
@@ -516,3 +529,67 @@ getLeaderSchedule globals ss cds poolHash key pp = Set.filter isLeader epochSlot
+      lvChainChecks = pparamsToChainChecksPParams . sgProtocolParams $ genesisShelley
+    }
+
+instance PraosCrypto c => ApplyTx (AllegraEra c)
```
Unlike the shelley instance, I moved the other era instances here, and kept the `PraosCrypto` constraint. Do we like this better?

nc6 replied:
No, this definitely seems wrong. `ApplyTx` has nothing to do with the protocol.

nc6 commented on eras/shelley-ma/impl/src/Cardano/Ledger/Allegra.hs:
```diff
@@ -39,12 +39,6 @@ import qualified Data.Map.Strict as Map
-
-instance PraosCrypto c => ApplyBlock (AllegraEra c)
-
-instance PraosCrypto c => GetLedgerView (AllegraEra c)
```
This one's a bit annoying. I am _tempted_ to say the correct thing to do with this is drop it entirely, and move it into consensus (in the way that I am doing with Praos). That would more accurately reflect its nature as an "integration" piece, rather than part of either ledger or protocol.

nc6 commented on eras/shelley-ma/impl/src/Cardano/Ledger/Allegra.hs:
```diff
@@ -39,12 +39,6 @@ import qualified Data.Map.Strict as Map
 
-instance PraosCrypto c => ApplyTx (AllegraEra c)
-
-instance PraosCrypto c => ApplyBlock (AllegraEra c)
```
Both `ApplyTx` and `ApplyBlock` are pure ledger concerns, right? `Block` actually calls `BBODY`. So these should stay ledger related.

nc6 commented on eras/shelley/impl/src/Cardano/Ledger/Shelley/API.hs:
```diff
@@ -6,49 +6,12 @@
-    Share (Core.TxOut era) ~ Interns (Credential 'Staking (Crypto era)),
-    FromSharedCBOR (Core.TxOut era)
-  ) =>
-  ShelleyBasedEra era
```
I think we should split this constraint block in two, rather than moving it. For one thing, as it is right now, it will not be satisfied once we move to Babbage anyway.

nc6 reviewed (changes requested):
Maybe let's discuss this in a call!

Commit:
Remove PraosCrypto from ShelleyEra

This involved adding:
DSignable crypto (Hash crypto EraIndependentTxBody)
to ShelleyEra in its place.

Commit:
remove the TPraos types from the Shelley API

Commit:
move Protocol API module to TPraos

Commit:
adjusting imports after API.Protocol moved

JaredCorduan commented:
@nc6 The plan worked out well!

nc6 reviewed (approved):
LGTM!
